File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool dfs(int curr,vector<int>& visited,vector<vector<int>>& adj){
        visited[curr]=1;
        for(int nei:adj[curr]){
            if(visited[nei]==0){
                if(!dfs(nei,visited,adj)){
                    return false;
                }
            }else if(visited[nei]==1){
                return false;
            }
        }
        visited[curr]=2;
        return true;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int n=numCourses;
        vector<vector<int>> adj(n);
        for(auto& p:prerequisites){
            adj[p[1]].push_back(p[0]);
        }
        vector<int> visited(n,0);
        for(int i=0;i<n;i++){
            if(visited[i]==0){
                if(!dfs(i,visited,adj)){
                    return false;
                }
            }
        }
        return true;
    }
};

File: 207-course-schedule/course-schedule.java
class Solution {
    public boolean canFinish(int n, int[][] edges) {
        ArrayList<ArrayList<Integer>> list=new ArrayList<>();
        for(int i=0;i<n;i++){
            list.add(new ArrayList<>());
        }
        for(int ed[]:edges){
            list.get(ed[0]).add(ed[1]);
        }
        Queue<Integer> q=new LinkedList<>();
        int visited[]=new int[n];
        for(int i=0;i<n;i++){
            if(visited[i]==0){
                if(dfs(visited,i,list)==false){
                    return false;
                }
            }
        }
        return true;
        
    }
    static boolean dfs(int[] visited,int node,ArrayList<ArrayList<Integer>> list){
        visited[node]=1;
        for(int nd:list.get(node)){
            if(visited[nd]==0){
                if(!dfs(visited,nd,list)){
                    return false;
                }
            }else if(visited[nd]==1){
                return false;
            }
        }
        visited[node]=2;
        return true;
    }
}